### ISSUE TITLE
Add CompetencyFramework to domain of Course (Hutchinson/Kansas)

### ISSUE THREAD

**Issue by stuartasutton** [CredentialEngine/Schema-Development#847]
Currently, the [ceterms:requires](https://credreg.net/ctdl/terms/requires) property on [ceterms:Course](https://credreg.net/ctdl/terms/Course) has a range of [ceterms:Competency](https://credreg.net/ctdlasn/terms#Competency). There is a a more frequent need for a missing range of [ceterms:CompetencyFramework](https://credreg.net/ctdlasn/terms#CompetencyFramework). For example, the Hutchinson (Kansas ) courses all have individual competency frameworks that are distinct from Program and College/University level competency frameworks.

**Comment by siuc-nate** [comment]
If this can reliably be interpreted as "it requires all of the competencies from the framework" (and I'm not sure what other practical interpretation there would be), then I don't think this should be an issue.

Worth thinking about: Does requiring a competency also (implicitly) require all of the children of that competency?  I think this notion had come up a few years back, but I don't remember the specifics. We did, I think, end up deciding that every individual required competency needed to be referenced, but I'm certainly open to revisiting that.

**Comment by philbarker** [comment]
At first sight that seemed odd, but taking the definition of `requires`: "Requirement or **set of requirements** for this resource." and the comment for `CompetencyFramework`: "A competency framework serves a container object for a **set of competencies**." I see how it should work. But I think Nate's right about what we should think about.

**Comment by stuartasutton** [comment]
I think my rationale was simple. When I create a course with learning outcomes specific to the course and embody those learning outcomes in a competency framework, I should be able to reference that framework as framing the individual outcomes of the course. If the desire of a deployment is to display all the competency individuals, then do that by resolving the individual competencies.

**Comment by siuc-nate** [comment]
I _think_ I get what you mean, but I feel like I'm missing something. So just to be sure:
- Would this be done instead of, or in addition to, referencing the competencies directly?
- Why is it not sufficient to use the existing connections (directly to competencies), as we have been doing all these years?
  - They are contextualized by their framework regardless, I would think, otherwise we would need to reexamine a lot of existing data as well

**Comment by stuartasutton** [comment]
No, I did not suggest dropping Competency as range. What I am struggling with is the “why not” of it. I can’t see how pointing to a framework that contains all and only the competencies this course addresses instead of forcing the itemization of each of competencies is a problem EXCEPT that we’ve hardwired itemization into our software.Sent from my iPhoneOn Sep 29, 2022, at 12:24 PM, siuc-nate ***@***.***> wrote:﻿
I think I get what you mean, but I feel like I'm missing something. So just to be sure:

Would this be done instead of, or in addition to, referencing the competencies directly?
Why is it not sufficient to use the existing connections (directly to competencies), as we have been doing all these years?

They are contextualized by their framework regardless, I would think, otherwise we would need to reexamine a lot of existing data as well

**Comment by siuc-nate** [comment]
I guess what I'm trying to get at is, are you wanting to point to the framework as a shortcut to pointing to its competencies, or because there is a particular semantic reason to point to the framework itself instead of to its competencies?  

- If you want a shortcut to pointing to the competencies, then the tooling should be updated to point to the competencies when the framework is specified, so you don't have to manually select/publish all of those competency URIs (the reference to the framework would not end up in the schema or the data, it would just be something extra in the API/tools to make it easier to publish)
- If you want to point to the framework instead of pointing to the competencies (or perhaps in parallel with doing so), then I'm trying to understand what the meaning/benefit of that is and why it's now an issue when it hasn't been one for the last several years. In other words, what does it _mean_ to point to the framework? What does that tell you that you don't already know from pointing to the competencies instead/as well?

What I want to avoid is a case where some data in the registry has things pointing to competencies, some data has things pointing to frameworks, some data has things pointing to both, and it's left up to the consumer of the data to figure out how to search for and display those things based on their references to competencies.  Pointing to the framework when you really mean pointing to its competencies adds a layer of indirection that compounds the existing array of hops a user already needs to take (both for searching and display), and I want to avoid incurring that unless there is a specific reason why you need to point to the framework independently of pointing to its competencies.

**Comment by philbarker** [comment]
@siuc-nate I think you and @stuartasutton can both achieve what you want if
1. we do as Stuart asks and add [ceterms:CompetencyFramework](https://credreg.net/ctdlasn/terms#CompetencyFramework) to the range of  [ceterms:requires](https://credreg.net/ctdl/terms/requires) 
2. you do as you want and modify the Registry tooling so that whenever a CompetencyFramework is seen as the value of ceterms:requires it is replaced with all the Competencies in the framework.

In other words the Registry profile of CTDL may be more restrictive than the full CTDL, if the tooling or use cases demand.

The risk in doing what you suggest, though, is that any updates to the Competency Framework after the data is ingested will not be reflected in the links from the Course to the Competencies. This maybe cuts both ways: the framework may change when the course doesn't.

**Comment by siuc-nate** [comment]
The thing is, if there is a different meaning behind pointing to the framework vs pointing to its competencies, then having that shortcut wouldn't be desired, since you'd be unable to **just** point to the framework if you wanted to do that for one reason or another.

That's why I'm trying to understand which of the two cases I outlined above is the one we're dealing with.

**Comment by stuartasutton** [comment]
@siuc-nate, if I have a course that requires a select set of competencies from a framework, I need to enumerate each of the members of that select set. However, if I have a framework that is designed specifically for a course (i.e., all competencies in the framework are required (and one could assume, assessed) and assumed downstream to be covered) then referencing the framework itself as required is all that is necessary.

**Comment by siuc-nate** [comment]
So the "correct" data would be direct references from the course to those competencies, then.  I agree with the utility of being able to just point to the framework instead from the publishing perspective, but what I'm wrestling with is how best to prevent that from creating a compounding problem for the search and display perspectives. As noted above, one potential solution would be to handle this in the publishing process itself, so that the actual data that ends up in the registry is a course that directly references those competencies (leaving the framework to be as transparent to the matter as it is today). If the competencies for that framework change, then this would require republishing the course to update those competency references - but that may be a rare enough thing, and have a minimal enough impact (particularly for courses that are 1:1 with their frameworks) to be an acceptable downside. I suppose we technically already have that problem today as it is, anyhow.

So, that would be (at the moment anyway - I'll keep chewing on this one) my proposed method of handling this:
- Don't make any change to the schema
- Update the various publishing tools to enable auto-selecting all of the competencies for a framework by selecting the framework itself
- Publish data that uses the same direct references to competencies that existing data uses today
- No need to alter anything with regard to search, display, handbook, etc.

**Comment by stuartasutton** [comment]
The reason I originally asked is because I am working on this NOW and planned to do more over the weekend. However, I am going to STOP and we can go over it synchronously on TUESDAY as opposed to rounds of posts here in github.

**Comment by siuc-nate** [comment]
Per our 2022-10-11 meeting:
- The reference to the set itself (of competencies), as a set, is semantically meaningful
- This set may be a Competency Framework or a Collection of competencies
- This may particularly be useful in cases where a Course, an Assessment, and/or a Credential all reference the same set of competencies (e.g. the Credential requires the same set of competencies that the Course teaches)
- We want to avoid breaking or impeding the technical implementation as it already exists, to the greatest extent possible
- Due to the volume of Competencies vs other kinds of things like required Courses, required Assessments, etc., this approach needs to focus on Competencies. We can consider the other types of objects later if needed (we don't want to create a solution that would impede that if we needed to go there someday).

We discussed two overall approaches:
- Extending the range of ceterms:teaches, ceterms:assesses, and ceterms:targetCompetency (not preferred)
- Creating new properties (e.g. ceterms:teachesAllFrom, ceterms:assessesAllFrom, ceterms:targetsAllCompetenciesFrom) (preferred)

To be determined:
- Whether the new properties might also reference a Competency directly where such a Competency has child Competencies, as a way of referencing that set of Competencies. This would be very useful, I think, as it avoids needing to create a Collection.

Other considerations:
- How to interpret a reference (via one of the above properties) to a Collection whose members include one or more other Collections where the top-level Collection and/or its member Collections contain competencies. Presumably this would just recursively gather all of the Competencies together as the things that are being taught/assessed/required, as if those Competencies were all in one Framework

The additional difficulty created for searches/consumers of the data is unfortunate, but unavoidable if we want to reference the set of things itself, which we do. In this situation, I currently believe it is less dangerous (in terms of technical implementation/impact) to mint new properties that explicitly reference a container of things than it is to extend existing properties to include a container. I will think about this some more and see if I become more (or less) certain about this stance. I would also like to hear from @mparsons-ce on this matter.

**Comment by siuc-nate** [comment]
Also consider that we have a very minimally-used https://credreg.net/ctdl/terms/targetCompetencyFramework property already, which seems relevant to this issue (good catch @mparsons-ce )

**Comment by siuc-nate** [comment]
This one may still be relevant for #947
